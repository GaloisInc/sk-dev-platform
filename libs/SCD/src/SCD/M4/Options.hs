{- |
Module      : $Header$
Description : Global options for reference policy processing
Copyright   : (c) Galois, Inc.

Global options for reference policy processing
-}

module SCD.M4.Options(Options(..), defaultOptions) where

-- | Global options for reference policy processing
data Options = Options{ implicitRequire :: Bool
                      , group           :: Bool
                      , xmlErrorOutput  :: Bool
                      , ifdefDeclFile   :: Maybe FilePath
                      }
  deriving Show

-- | Default options for reference policy processing
defaultOptions :: Options
defaultOptions = Options{ implicitRequire = False
                        , group           = True 
                        , xmlErrorOutput  = False
                        , ifdefDeclFile   = Nothing
                        }

