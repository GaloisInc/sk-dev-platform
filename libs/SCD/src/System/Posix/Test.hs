import System.Posix.Interaction(Interaction(..), OStream(..), testProgram)

-- throw-away example demonstrating Interactions

main :: IO ()
main = testProgram "/usr/bin/head" ["-1"] Nothing True (Interactions [Feed "apa\n", Expect (Just 2) StdOut "apa\n"])
