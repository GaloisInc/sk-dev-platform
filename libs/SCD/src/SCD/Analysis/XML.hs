{-# LANGUAGE ScopedTypeVariables #-}
--module SCD.Analysis.XML where

import Text.XML.Light(parseXML, showElement, strContent, elChildren,
  unqual, Content(..), Element(..), Attr(..), cdData, Line)

import Text.Regex.Posix((=~))
import Data.Char(isSpace)

import Control.Monad(when)
import System.Environment(getArgs)

import Data.List.GroupSort(groupSort)

import Text.ParserCombinators.PolyReadP(ReadP, parse, option,
  optional, many, get, munch)

import Text.PrettyPrint.HughesPJ(render, nest, ($+$), text, empty, Doc, (<+>),(<>),parens,colon, sep)

data InterfaceModule = InterfaceModule{ file :: String
                                      , moduleSummary :: String
                                      , moduleDescription :: [Content]
                                      , required :: Bool
                                      , requiredDescription :: String
                                      , interfaces :: [Interface]
                                      }
  deriving Show

data Interface = Interface{ interfaceSummary :: String
                          , interfaceDescription :: [Content]
                          , parameters :: [Parameter]
                          , infoFlow :: Maybe InfoFlow
                          }
  deriving Show

data InfoFlow = InfoFlow{ flow :: Flow
                        , weight :: Int
                        } 
  deriving (Eq, Read, Show)

data Flow = Read | Write | Both | None
  deriving (Eq, Read, Show)


data Parameter = Parameter{ line :: Maybe Line
                          , name :: String
                          , parameterSummary :: String
                          }
  deriving (Eq, Read, Show)

type P a = ReadP [Content] Content a

withContent :: Show a => [Content] -> P a -> P a
withContent s p = either (\e -> fail e) return $ parse p s

parseFile :: String -> IO (Either String InterfaceModule)
parseFile f = do
  s <- readFile f
  return (parseModule f s)

main :: IO ()
main = do
   ca:fs <- getArgs
   c <- case ca of "--per-parameter" -> return collect
                   "--all-parameters" -> return collectAllParams
                   _                  -> fail $ "Unknown flag: " ++ show ca
   mapM (\f -> parseFile f >>= either (fail.show.show) return) fs >>= (putStrLn . c)

above :: [Doc] -> Doc
above = foldr ($+$) empty

collect :: [InterfaceModule] -> String
collect ms = render $ above
           $ map (\(n,l) -> text "******" <+> text n $+$ 
                             (nest 4 $ above $ 
                                map (\(s,l') -> 
                                     text (unwords (lines s)) $+$ nest (-4) (
                                     above ([text (file m) <> colon <> text (show (line p)) <> colon <+> text n <+> parens (text (unwords (lines s)))
                                                           | (m,_i,p) <- l'])))
                                (groupSort (\(_m,_i,p)->
                                              dropWhile isSpace $
                                              parameterSummary p) l)))
           $ groupSort (\(_m,_i,p) -> name p)
           $ concat (concat [[[(m,i,p) | p <- parameters i] 
                                  | i <- interfaces m] 
                             | m <- ms])

collectAllParams :: [InterfaceModule] -> String
collectAllParams ms = render $ above
           $ map (\(ns,l) -> let dns = sep (map text ns) in 
                             text "******" <+> dns $+$ 
                             (above $ 
                                map (\(m,i) -> text (file m) <> colon <> text (show (line( head (parameters i)))) <> colon <+> text (unwords (lines (interfaceSummary i)))) l))
           $ groupSort (\(_m,i) -> map name (parameters i))
           $ concat ([[(m,i) | i <- interfaces m] 
                             | m <- ms])

parseModule :: String -> String -> Either String InterfaceModule
parseModule filename = parse (parseModule' filename) . parseXML 
                     . extractXML

-- from refpolicy/support/segenxml.py: XML_COMMENT = re.compile("^##\s+(.*?)\s*$")

extractXML :: String -> String
extractXML l = unlines [after | (_before::String,_pattern::String,after) <- map (=~ "^##\\s") (lines l)]

parseModule' :: String -> P InterfaceModule
parseModule' filename = do
  s <- parseSummary
  d <- parseDescription
  (r,rd) <- option (False,"") parseRequired
  is <- many parseInterface
  return InterfaceModule{ file = filename
                        , moduleSummary = s
                        , moduleDescription = d
                        , required = r
                        , requiredDescription = rd
                        , interfaces = is
                        } 

parseSummary :: P String
parseSummary = do
  s <- parseElement "summary"
  when (not (null (elChildren s))) $ 
       fail $ "Illegal content in summary: " ++ showElement s
  return (strContent s)

parseDescription :: P [Content]
parseDescription = option [] (elContent `fmap` parseElement "desc")
      
parseInterface :: P Interface
parseInterface = do
  s <- parseSummary
  d <- parseDescription
  ps <- many parseParameter
  i <- option Nothing (Just `fmap` parseInfoFlow)
  optional (parseElement "rolecap")
  optional (parseElement "rolebase")
  return Interface{ interfaceSummary = s
                  , interfaceDescription = d
                  , parameters = ps
                  , infoFlow = i
                  }

parseInfoFlow :: P InfoFlow
parseInfoFlow = do
  p <- parseElement "infoflow"
  fs <- parseAttr "type" p
  f <- case fs of
         "read"  -> return Read
         "write" -> return Write
         "none"  -> return None
         "both"  -> return Both
         _       -> fail $ "Illegal infoflow type: " ++ show fs
  ws <- parseAttr "weight" p
  w <- case reads ws of 
         [(a,"")] -> return a
         _        -> fail $ "Illegal infoflow weight: " ++ show ws
  return InfoFlow{ flow = f, weight = w }

parseParameter :: P Parameter
parseParameter = do
  p <- parseElement "param"
  n <- parseAttr "name" p
  s <- withContent (elContent p) parseSummary 
  return Parameter{ line = elLine p
                  , name = n
                  , parameterSummary = s
                  }

parseRequired :: P (Bool,String)
parseRequired = do
  p <- parseElement "required"
  n <- parseAttr "val" p
  r <- case n of
         "true"  -> return True
         "false" -> return False
         _       -> fail $ "Illegal boolean: " ++ show n
  return (r, strContent p)
  
parseElement :: String -> P Element
parseElement t = do
  eatWhiteSpace
  e <- tok (\c -> 
             case c of Elem e | elName e == unqual t -> Just e
                       _ -> Nothing)
  eatWhiteSpace
  return e

eatWhiteSpace :: P [Content]
eatWhiteSpace = munch $ \c ->
                case c of 
                  Text cd | all isSpace (cdData cd) -> True
                  _ -> False

parseAttr :: String -> Element -> P String
parseAttr t e = case [attrVal a | a <- elAttribs e, attrKey a == unqual t] of
                  v:_ -> return v
                  []  -> fail $ "Expected attribute "++show t++" at "++
                                showElement e

tok :: (Content -> Maybe a) -> P a
tok p = get >>= (maybe (fail "token") return . p)
