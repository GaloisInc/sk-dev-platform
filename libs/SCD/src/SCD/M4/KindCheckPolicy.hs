{- |
Module      : $Header$
Description : Entry point for kind-checking SELinux policies.
Copyright   : (c) Galois, Inc.

This module provides an entry point for parsing and subsequently
kind-checking an SELinux policy directory.  It currently supports error
output in human-readable, textual form, or in XML.
-}

module SCD.M4.KindCheckPolicy(kcPolicyDoc, errorDoc, xmlErrorDoc) where

import SCD.M4.Options(Options(..))
import SCD.M4.KindCheck(KindCheckOptions(..), kcPolicy)
import SCD.M4.Errors(Error, occurrences, flatten, errorsByType)
import SCD.M4.ErrorsXml(genXmlOutput)
import SCD.M4.ModuleFiles(readPolicy, readPolicyModule)
import Text.PrettyPrint.HughesPJ(Doc, empty, text, ($+$))
import Text.PrettyPrint.Pp(above, pp, pnest)
import Text.XML.Light(Element)
import Data.List(sort, intersperse)
import System.Directory(canonicalizePath)

-- | Parses and kind-checks the given SELinux policy directory.  Options
-- can control many aspects of the kind-checking behavior (e.g., XML
-- vs. textual error output, etc).
kcPolicyDoc :: Options -> KindCheckOptions -> FilePath -> [FilePath] -> IO (Doc, Maybe Element)
kcPolicyDoc os kos pd pmfs = do
  pd':pmfs' <- if xmlErrorOutput os then mapM canonicalizePath (pd:pmfs)
                                   else return (pd:pmfs)
  pms <- mapM readPolicyModule pmfs'
  p <- readPolicy (ifdefDeclFile os) pd'
  let (_ke, errs) = kcPolicy os kos p pms
  let edoc = if xmlErrorOutput os then xmlErrorDoc else errorDoc
  return $ edoc os errs

-- | Generates textual error output for the given shrimp errors.
errorDoc :: Options -> [Error] -> (Doc, Maybe Element)
errorDoc os errs =
    (text "*** Statistics:" $+$
     above (intersperse (text "") (
         [ pp e $+$ if l > 1 then pnest (text ("("++show l++" errors like this.)")) else empty
         | (l,e) <- reverse (sort (occurrences (flatten errs)))])) $+$
     text "" $+$
     text "*** Complete error list:" $+$
     above (intersperse (text "") (if group os then concat (intersperse [text "--"]
                                                            (map (map pp) (errorsByType errs)))
                                   else map pp errs)),
     Nothing)                                   

-- | Generates XML error output for the given shrimp errors.  Note that
-- the XML parse tree will be Nothing when there are no errors present.
xmlErrorDoc :: Options -> [Error] -> (Doc, Maybe Element)
xmlErrorDoc _os errs = (text t, e) where (t,e) = genXmlOutput errs
