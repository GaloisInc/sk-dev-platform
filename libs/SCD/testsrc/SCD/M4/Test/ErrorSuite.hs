{- |
Module      : $Header$
Description : Tests for Shrimp M4 error suite
Copyright   : (c) Galois, Inc.

Tests for Shrimp M4 error suite.  The error suite is an extremely
lightweight version of the reference policy that raises all of the
errors that Shrimp is capable of raising.  This module parses and
kind-checks the error suite policy, and then compares the resulting
parse tree of XML errors versus a golden XML file from the repository.
-}

module SCD.M4.Test.ErrorSuite (checks, testCases) where

import SCD.M4.KindCheck          (defaultKindCheckOptions)
import SCD.M4.KindCheckPolicy    (kcPolicyDoc)
import SCD.M4.ErrorsXml          (markerQN, fileQN)
import SCD.M4.Options            (Options(..), defaultOptions)
import System.Directory          (getCurrentDirectory)
import System.FilePath           (splitPath, joinPath, dropTrailingPathSeparator, (</>))
import Control.Arrow             (second)
import Text.Regex                (mkRegex, subRegex)
import Text.XML.Light            (Content(..), Element(..), Attr(..), CData(..), parseXMLDoc, ppTopElement)

import Test.Framework.Providers.HUnit
import Test.Framework.Providers.API ( Test )
import Test.HUnit hiding ( Test )

testCases :: FilePath -> [Test]
testCases path = [testCase "Golden xml comparison" checkXml]
    where checkXml = do
            cwd <- getCurrentDirectory
            let pd         = cwd </> path
                opts       = defaultOptions{ xmlErrorOutput = True }
                process    = procXML . parseXMLDoc . cleanup

            (_,xml) <- second procXML `fmap` kcPolicyDoc opts defaultKindCheckOptions pd []
            refXml  <- process `fmap` readFile (pd </> "ErrorSuite.golden.xml")

            let render = maybe "" ppTopElement
                failMsg = "FAILED:\n" ++ render xml ++ "\nGolden:\n" ++ render refXml++"\n"

            writeFile "xml.out"   $ render xml
            writeFile "refXml.out" $ render refXml

            -- Compare the parse tree of the golden XML output (after prepending
            -- the appropriate parent path) against the parse tree of the
            -- just-generated XML.  They should be exactly the same.

            assertEqual failMsg xml refXml


-- | Verify the resulting errors from parsing and kind-checking the
-- error suite policy versus the \"golden\" XML file.
checks :: FilePath -> IO ()
checks path = do
  putStrLn "Begin tests for shrimp error generation (ErrorSuite)"

  cwd <- getCurrentDirectory
  let pd         = cwd </> path
      opts       = defaultOptions{ xmlErrorOutput = True }
      process    = procXML . parseXMLDoc . cleanup

  (_,xml) <- second procXML `fmap` kcPolicyDoc opts defaultKindCheckOptions pd []
  refXml  <- process `fmap` readFile (pd </> "ErrorSuite.golden.xml")

  let render = maybe "" ppTopElement
  --writeFile "newxml.out"   $ render xml
  --writeFile "transxml.out" $ render refXml

  -- Compare the parse tree of the golden XML output (after prepending
  -- the appropriate parent path) against the parse tree of the
  -- just-generated XML.  They should be exactly the same.

  putStrLn $ "Shrimp error suite " ++
    if xml == refXml then "passed." else
    "FAILED:\n" ++ render xml ++ "\nGolden:\n" ++ render refXml++"\n"
  putStrLn "End tests for shrimp error generation (ErrorSuite)"
  return ()

-- For some reason, the XML parser considers whitespace characters as
-- part of the CDATA content of the parsed XML, which throws off parse
-- tree comparision.  This function strips the extraneous whitespace.
cleanup :: String -> String
cleanup s = subRegex (mkRegex ">[[:space:]]*<") s "><"

prefix :: String
prefix = "ErrorSuite"

-- | Deletes irrelevant prefix of path to each marker element's file
-- | attribute, and rips out line information.
procXML :: Maybe Element -> Maybe Element
procXML Nothing  = Nothing
procXML (Just e) = Just $ e { elContent = map procXML' (elContent e)
                            , elLine = Nothing
                            }

procXML' :: Content -> Content
procXML' (Elem el) =
  Elem $ el{ elAttribs = attrs
           , elContent = children
           , elLine    = Nothing
           }
  where
    children :: [Content]
    children = map procXML' (elContent el)

    attrs :: [Attr]
    attrs | (elName el == markerQN) = map fixAttr (elAttribs el)
          | otherwise               = elAttribs el

    fixAttr :: Attr -> Attr
    fixAttr a | (attrKey a == fileQN) = a{ attrVal = fixPath (attrVal a) }
              | otherwise = a

    fixPath :: FilePath -> FilePath
    fixPath = joinPath . dropWhile ((/=prefix) . dropTrailingPathSeparator) . splitPath

procXML' (Text cd) = Text $ cd{cdLine = Nothing}
procXML' c         = c
