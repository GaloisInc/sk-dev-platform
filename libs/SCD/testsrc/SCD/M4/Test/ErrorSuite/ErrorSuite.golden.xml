<?xml version='1.0' ?>
<shrimp>
  <warnings>
    <duplicate_access_vector_def>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/access_vectors" name="error_class_one" line="115" col="7" start_offset="1534" end_offset="1549" />
      <defined_at>
        <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/access_vectors" name="error_class_one" line="114" col="7" start_offset="1497" end_offset="1512" />
      </defined_at>
    </duplicate_access_vector_def>
    <duplicate_class_permission>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/access_vectors" name="read" line="120" col="32" start_offset="1706" end_offset="1710" />
    </duplicate_class_permission>
    <duplicate_common_def>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/access_vectors" name="error_common_def" line="10" col="8" start_offset="255" end_offset="271" />
      <defined_at>
        <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/access_vectors" name="error_common_def" line="9" col="8" start_offset="216" end_offset="232" />
      </defined_at>
    </duplicate_common_def>
    <duplicate_common_permission>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/access_vectors" name="read" line="15" col="36" start_offset="442" end_offset="446" />
    </duplicate_common_permission>
    <duplicate_definitions>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="error_duplicate_def" line="68" col="12" start_offset="1764" end_offset="1783" />
    </duplicate_definitions>
    <in_implementation layer="test" module="errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errorstwo.te" name="errors" line="6" col="15" start_offset="290" end_offset="296" />
      <warnings>
        <duplicate_identifier>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errorstwo.te" name="errors" line="6" col="15" start_offset="290" end_offset="296" />
          <defined_at>
            <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="errors" line="1" col="15" start_offset="14" end_offset="20" />
          </defined_at>
        </duplicate_identifier>
      </warnings>
    </in_implementation>
    <duplicate_macro_def>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/obj_perm_sets.spt" name="dup_macro_def" line="10" col="9" start_offset="325" end_offset="338" />
      <defined_at>
        <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/obj_perm_sets.spt" name="dup_macro_def" line="9" col="9" start_offset="283" end_offset="296" />
      </defined_at>
    </duplicate_macro_def>
    <in_implementation layer="test" module="errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="errors" line="1" col="15" start_offset="14" end_offset="20" />
      <warnings>
        <duplicate_symbol_decl>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="duplicate_symbol_decl_t" line="7" col="6" start_offset="113" end_offset="136" />
          <defined_at>
            <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="duplicate_symbol_decl_t" line="6" col="6" start_offset="83" end_offset="106" />
          </defined_at>
        </duplicate_symbol_decl>
        <duplicate_symbol_decl>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="duplicate_identifier_t" line="12" col="11" start_offset="206" end_offset="228" />
          <defined_at>
            <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="duplicate_identifier_t" line="11" col="11" start_offset="172" end_offset="194" />
          </defined_at>
        </duplicate_symbol_decl>
      </warnings>
    </in_implementation>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <fragment_kind_error>
          <expected_kind>
            <kind name="domain_" />
          </expected_kind>
          <got_complex_param>
            <complex_param_comp value="-undefined_type">
              <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="undefined_type" line="33" col="34" start_offset="811" end_offset="825" />
            </complex_param_comp>
          </got_complex_param>
        </fragment_kind_error>
      </warnings>
    </in_definition>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <illegal_fragment>
          <fragment_text value="domain_trans" />
          <defined_at>
            <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/misc_patterns.spt" name="$1domain_trans" line="11" col="35" start_offset="333" end_offset="347" />
          </defined_at>
        </illegal_fragment>
      </warnings>
    </in_definition>
    <in_implementation layer="test" module="errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="errors" line="1" col="15" start_offset="14" end_offset="20" />
      <warnings>
        <illegal_macro_use>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="domain_trans" line="16" col="6" start_offset="307" end_offset="319" />
          <defined_at>
            <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/misc_patterns.spt" name="domain_trans" line="24" col="9" start_offset="712" end_offset="724" />
          </defined_at>
        </illegal_macro_use>
      </warnings>
    </in_implementation>
    <in_implementation layer="test" module="errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="errors" line="1" col="15" start_offset="14" end_offset="20" />
      <warnings>
        <illegal_parameter_use>
          <illegal_parameter index="1">
            <kind name="type_" />
          </illegal_parameter>
        </illegal_parameter_use>
      </warnings>
    </in_implementation>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <illegal_symbol_decls>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="$1" line="46" col="14" start_offset="1297" end_offset="1299" />
        </illegal_symbol_decls>
      </warnings>
    </in_definition>
    <in_definition macro_name="another_placeholder_interface">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errorsthree.if" name="another_placeholder_interface" line="18" col="12" start_offset="325" end_offset="354" />
      <warnings>
        <illegal_symbol_ref>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errorsthree.if" name="valid_type_t" line="24" col="22" start_offset="554" end_offset="566" />
          <defined_at>
            <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="valid_type_t" line="3" col="6" start_offset="34" end_offset="46" />
          </defined_at>
        </illegal_symbol_ref>
      </warnings>
    </in_definition>
    <missing_access_vector_def>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/security_classes" name="error_class_two" line="6" col="7" start_offset="77" end_offset="92" />
    </missing_access_vector_def>
    <in_implementation layer="test" module="errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errorstwo.te" name="errors" line="6" col="15" start_offset="290" end_offset="296" />
      <warnings>
        <module_id_mismatch>
          <file_base_name name="errorstwo" />
          <module_name name="errors" />
        </module_id_mismatch>
      </warnings>
    </in_implementation>
    <mutually_recursive>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="mutrec_iface" line="8" col="12" start_offset="190" end_offset="202" />
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errorsthree.if" name="mutrec_other_iface" line="9" col="12" start_offset="163" end_offset="181" />
    </mutually_recursive>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <ref_policy_warn_call>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="stat_file_perms" line="49" col="59" start_offset="1394" end_offset="1409" />
          <warning_text>$0 is deprecated please use getattr_file_perms instead.</warning_text>
        </ref_policy_warn_call>
      </warnings>
    </in_definition>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <kind_mismatch>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="file_class_set" line="29" col="43" start_offset="718" end_offset="732" />
          <expected_kind>
            <kind name="permission" />
          </expected_kind>
          <got_kinds>
            <kind name="class" />
            <kind name="sid" />
          </got_kinds>
        </kind_mismatch>
      </warnings>
    </in_definition>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <undefined_macro>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="this_macro_is_undefined" line="52" col="9" start_offset="1450" end_offset="1473" />
        </undefined_macro>
      </warnings>
    </in_definition>
    <undefined_common_id>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/flask/access_vectors" name="invalid_common_name" line="123" col="33" start_offset="1774" end_offset="1793" />
    </undefined_common_id>
    <undefined_identifier>
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/obj_perm_sets.spt" name="undefined_perms" line="12" col="40" start_offset="399" end_offset="414" />
    </undefined_identifier>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <undefined_identifier>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="undefined_t" line="42" col="27" start_offset="1151" end_offset="1162" />
        </undefined_identifier>
        <undefined_identifier>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="undefined_type_t" line="24" col="22" start_offset="517" end_offset="533" />
        </undefined_identifier>
      </warnings>
    </in_definition>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <undefined_identifiers>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="undefined_t" line="42" col="27" start_offset="1151" end_offset="1162">
            <kind name="type/attribute" />
          </marker>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="undefined_t" line="42" col="15" start_offset="1139" end_offset="1150">
            <kind name="domain" />
          </marker>
        </undefined_identifiers>
      </warnings>
    </in_definition>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <undocumented_parameters>
          <param_info optional="false">
            <kind name="type" />
          </param_info>
        </undocumented_parameters>
      </warnings>
    </in_definition>
    <in_implementation layer="test" module="errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="errors" line="1" col="15" start_offset="14" end_offset="20" />
      <warnings>
        <unknown_ifdef_id>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.te" name="unknown_symbol" line="22" col="8" start_offset="384" end_offset="398" />
        </unknown_ifdef_id>
      </warnings>
    </in_implementation>
    <in_definition macro_name="domain_unused_parameters">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/misc_patterns.spt" name="domain_unused_parameters" line="14" col="9" start_offset="433" end_offset="457" />
      <warnings>
        <unused_parameters>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/misc_patterns.spt" name="domain_unused_parameters" line="14" col="9" start_offset="433" end_offset="457" />
          <param_info optional="false">
            <kind name="type/attribute" />
            <kind name="domain" />
          </param_info>
          <param_info optional="false">
            <kind name="any" />
          </param_info>
          <param_info optional="false">
            <kind name="type/attribute" />
          </param_info>
        </unused_parameters>
      </warnings>
    </in_definition>
    <in_definition macro_name="various_errors">
      <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="various_errors" line="21" col="12" start_offset="425" end_offset="439" />
      <warnings>
        <wrong_number_of_arguments>
          <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/modules/test/errors.if" name="domain_trans" line="39" col="9" start_offset="1044" end_offset="1056">
            <param_info optional="false">
              <kind name="domain" />
            </param_info>
            <param_info optional="false">
              <kind name="type/attribute" />
            </param_info>
            <param_info optional="false">
              <kind name="type/attribute" />
            </param_info>
          </marker>
          <defined_at>
            <marker file="/Users/jstanley/svn/scd/SCD/src/SCD/M4/Test/ErrorSuite/policy/support/misc_patterns.spt" name="domain_trans" line="24" col="9" start_offset="712" end_offset="724" />
          </defined_at>
        </wrong_number_of_arguments>
      </warnings>
    </in_definition>
  </warnings>
</shrimp>
